EU-Bound Shipping Notice: Create Top Banner

ThomazFB opened the pull request:
Part of #9591 

Why
==========
Since March 1, 2023, the EU has created new rules for customs declaration when shipping. This PR introduces a Notice warning the user about these new EU constraints.

How
==========
Configures a custom `TopBannerView` as the header of the `ShippingLabelFormViewController` tableView. This banner is created through a Factory, and it's visibility is currently controlled by the `.euShippingNotification` Feature Flag.

⚠️ The dismiss action is not being persisted right now, this will be fixed in #9695 

⚠️ The notice follows the UI guideline of the `TopBannerView` component, which deviates from the Figma suggestion. I verified that multiple banners are presented this way in different app areas, so I followed it instead of modifying anything to match Figma.

<details>
  <summary>Examples</summary>
  
| Order details notice  | Order list notice |
| ------------- | ------------- |
| ![Simulator Screen Shot - iPhone 14 Pro - 2023-05-10 at 19 50 04](https://github.com/woocommerce/woocommerce-ios/assets/5920403/e75934b4-7512-40c8-83e0-a11b219d5e64) | ![Simulator Screen Shot - iPhone 14 Pro - 2023-05-10 at 19 50 07](https://github.com/woocommerce/woocommerce-ios/assets/5920403/8f6d4a77-6309-49d3-b135-0f16ff299f01) |
</details>

Screen Capture
==========
https://github.com/woocommerce/woocommerce-ios/assets/5920403/023b4b05-8ed8-4822-8abf-a9f826167f50

How to Test
==========
### Scenario 1
1. Open the app with a site containing the Shipping Labels plugin configured
2. Open the order details of an order with the `processing` status
3. Hit the `Create Shipping Label` button
4. Verify that the Banner is correctly displayed inside the view
5. Hit the `Learn more` button and verify that a web view with the `USPS` page explaining the new Customs rule opens
6. Verify that hitting the dismiss button works

### Scenario 2
1. Disable the `.euShippingNotification` feature flag
2. Open the app with a site containing the Shipping Labels plugin configured
3. Open the order details of an order with the `processing` status
4. Hit the `Create Shipping Label` button
5. Verify that the Banner DOES NOT show up inside the view

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Define ShippingLabelsEUNoticeTopBannerFactory.swift type

ThomazFB committed:
Configure basic TopBannerView creation inside ShippingLabelsEUNoticeTopBannerFactory.swift

ThomazFB committed:
Configure basic implementation of the TopBannerView creation inside ShippingLabelFormViewController.swift

ThomazFB committed:
Fix ShippingLabelsEUNoticeTopBannerFactory build issues

ThomazFB committed:
Connect top banner call inside ShippingLabelFormViewController.swift

ThomazFB committed:
Merge branch 'trunk' into issue/9591-create-eu-bound-notice

ThomazFB committed:
Adjust ShippingLabelsEUNoticeTopBannerFactory.swift to correctly create action buttons

ThomazFB committed:
Update ShippingLabelsEUNoticeTopBannerFactory.swift strings

ThomazFB committed:
Adjust ShippingLabelFormViewController.swift to correctly insert the tableView header

ThomazFB committed:
Refactor and rename EU Shipping banner factory

ThomazFB committed:
Update EUShippingNoticeTopBannerFactory.swift UI configuration

ThomazFB committed:
Update ShippingLabelFormViewModel.swift to control when the banner should be visible

ThomazFB committed:
Configure learn more handling to EUShippingNoticeTopBannerFactory.swift

ThomazFB committed:
Refactor EUShippingNoticeTopBannerFactory.swift to send URL instead of String when learn more is clicked

ThomazFB committed:
Handle learn more click inside ShippingLabelFormViewController.swift

ThomazFB committed:
Add dismiss action handling to EUShippingNoticeTopBannerFactory.swift

ThomazFB committed:
Handle dismiss press event in Shipping notice top banner

ThomazFB committed:
Refactor banner handling code in ShippingLabelFormViewController.swift

ThomazFB committed:
Fix hound issues

ThomazFB committed:
Add documentation to the EU Shipping Notice functions and values

ThomazFB committed:
Merge branch 'trunk' into issue/9591-create-eu-bound-notice

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/EUShippingNoticeTopBannerFactory.swift:
```diff
@@ -0,0 +1,40 @@
+
+    enum String {
+        static let shippingCustomsInstructionsForEUCountries = "https://www.usps.com/international/new-eu-customs-rules.htm"
+    }
```
I'd suggest adding this to the list of URL constants in `WooConstants.URLs`, which is a single source of trusted URL string constants used in the app.

Also, that way the URL also doesn't have to be a parameter in the `onLearnMorePressed` closure. Instead, you can use the constant directly in `ShippingLabelFormViewController.presentShippingInstructionsView()`.

ThomazFB replied:
Oh, I see! I considered adding it to the `WooConstants` because of the safe `asURL` declaration. Still, I got confused as I noticed that all URLs there are Automattic related and decided to do something like what I saw in the `FancyAlertViewController+UnifiedLogin`, but keeping a unified source of URLs is much better indeed!

Fixed in https://github.com/woocommerce/woocommerce-ios/pull/9636/commits/e0c8ab06f1972fac0dccb4361db4a269ae5a4537 and https://github.com/woocommerce/woocommerce-ios/pull/9636/commits/9a70383a5ee39febb5d61e50574048b5f563e7a5

Thanks!

rachelmcr replied:
> I noticed that all URLs there are Automattic related

That's true, but my understanding is that "trusted" in this case is "trusted to be well-formed" (rather than "internal to Automattic) because when we call `asURL()` the app will crash if the URL is malformed. This is in contrast to non-constants that we're getting on the fly and can't trust to be well-formed URLs.

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift:
```diff
@@ -546,6 +556,55 @@ private extension ShippingLabelFormViewController {
+        configuration.secureInteraction = true
+        let webKitVC = WebKitViewController(configuration: configuration)
+        let nc = WooNavigationController(rootViewController: webKitVC)
+        self.present(nc, animated: true)
```
One option to simplify this (and for consistency with how we do this in other banners) is to use `WebviewHelper` to launch the webview. That way, all you need is to pass the URL string into `WebviewHelper.launch(_:with:)` and the helper does the rest of the work.

ThomazFB replied:
Nice!! Much better! Thanks for letting me know about this useful helper 😄. Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/9636/commits/bbd4d4ce02c16e9b067e9a54e68341e1aaace717

rachelmcr reviewed (commented):
Looking good. 👍  I have a couple suggestions to help simplify how the banner is created and launched; let me know what you think!

I also noticed that the text in the banner is quite small, but from what I'm seeing that seems to be a limitation of how we're handling the info text in `TopBannerView` generally. It might be worth investigating that in case we can make the text a more reasonable size (and have it respond to dynamic font sizes).

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift:
```diff
@@ -546,6 +556,55 @@ private extension ShippingLabelFormViewController {
+        }
+
+        let topBannerView = EUShippingNoticeTopBannerFactory.createTopBanner(
+            onDismissPressed: {
+                self.hideTopBannerView()
+            },
+            onLearnMorePressed: { instructionsURL in
+                self.presentShippingInstructionsView(instructionsURL: instructionsURL)
+            })
```
Since these are `@escaping` closures, it's safer to capture `self` weakly to avoid possible retain cycles.

ThomazFB replied:
Oh, true! I keep forgetting about this 😅, thanks! Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/9636/commits/1069ba58b8a9328af623621a3c2ead17896b83c7

ThomazFB committed:
Add USPS URL to WooConstants.swift

ThomazFB committed:
Remove URL parameter coming from EUShippingNoticeTopBannerFactory.swift

ThomazFB committed:
Adapt usage of WebviewHelper in ShippingLabelFormViewController.swift

ThomazFB committed:
Fix incorrect URL placement in WebviewHelper call

ThomazFB committed:
Weakly capture ShippingLabelFormViewController.swift

ThomazFB commented:
Hey @rachelmcr! Thank you for your great review! I've addressed all your suggestions in the comments and the PR is ready for a second round!

>I also noticed that the text in the banner is quite small, but from what I'm seeing that seems to be a limitation of how we're handling the info text in TopBannerView generally. It might be worth investigating that in case we can make the text a more reasonable size (and have it respond to dynamic font sizes).

Yeah, I find it to be kinda weird the way the banner is presenting its content, but since the component is used in so many parts of the app I wasn't really sure if this was intentional or not. I can try to fix that as a nice-to-have task from the project, WDYT?

rachelmcr reviewed (approved):
Changes look great! :shipit: 

> since the component is used in so many parts of the app I wasn't really sure if this was intentional or not. I can try to fix that as a nice-to-have task from the project, WDYT?

I think in most places the banner is used, the text isn't long enough for it to be an issue (except if you want dynamic text sizes). It probably just hasn't been investigated before — a nice-to-have task sounds perfect!
